src/lib/sqlite/wa-sqlite.ts

```diff
@@ -0,0 +1,265 @@
+  async function executeFn(query: string, ...params: SQLiteCompatibleType[]): Promise<SQLiteCompatibleType[][]> {
+    const rows = []
+    for await (const stmt of sqlite3.statements(db, query)) {
+      params.forEach((param, index) => sqlite3.bind(stmt, index + 1, param))
+      while (await sqlite3.step(stmt) === SQLite.SQLITE_ROW) {
+        rows.push(sqlite3.row(stmt))
+      }
+    }
+    return rows
+  }
+  async function close() {
+    await sqlite3.close(db)
+  }
+  async function executeLocking(query: string, ...params: SQLiteCompatibleType[]) {
+    return await navigator.locks.request('crypta_executor', (lock) => executeFn(query, ...params))
+  }
+  async function executeWithoutLocking(query: string, ...params: SQLiteCompatibleType[]) {
+    return await executeFn(query, ...params)
+  }
+
+  if (locking) {
+    return {
+      execute: executeLocking,
+      close,
+    }
+  } else {
+    return {
+      execute: executeWithoutLocking,
+      close,
+    }
+  }
+}
+
+export function createLocalDb(executor: WASqliteExecutor) {
+  return drizzle(async (queryString, params, method) => {
+    const result = await executor.execute(queryString, ...params)
+    if (method === 'get' && result.length > 0) {
+      return {rows: result[0]}
+    }
+    return {rows: result}
+  })
+}
+
+export async function migrateLocal(
+  executor: WASqliteExecutor,
+  dataImportFn: (executor: WASqliteExecutor) => Promise<void>,
+  migrationQueryMap: MigrationQueryMap,
+  queriesStringMap: QueriesStringMap,
+) {
+  let [[currentUserVersion]] = await executor.execute('PRAGMA user_version') as [[number]]
+  while (migrationQueryMap[currentUserVersion]) {
+    const migrationQueryPath = migrationQueryMap[currentUserVersion]
+    const migrationQueryString = queriesStringMap[migrationQueryPath]
+    if (!migrationQueryString) {
+      throw new Error(`migrate: could not find query string of ${migrationQueryPath}`)
+    }
+    await executor.execute(migrationQueryString)
+    if (currentUserVersion === 0) {
+      await dataImportFn(executor)
+      const path = '/db/0000_seed_default_folder.sql'
+      const seedFolderQuery = queriesStringMap[path]
+      await executor.execute(seedFolderQuery)
+    }
+    currentUserVersion += 1
+    await executor.execute(`PRAGMA user_version = ${currentUserVersion}`)
+  }
+}
+
+export async function createLocalSnippetsStore(
+  executor: WASqliteExecutor,
+  globalStateStore: Writable<GlobalState>,
+): Promise<SnippetStore> {
+  let snippets: Snippet[] = []
+  let folderId = 'default'
+  const db = createLocalDb(executor)
+  const store = writable(snippets)
+  const migrationStateStore = writable<MigrationState>('not-started')
+  try {
+    migrationStateStore.set('running')
+    await migrateLocal(
+      executor,
+      async () => {},
+      defaultMigrationQueryMap,
+      defaultQueriesStringMap,
+    )
+    migrationStateStore.set('done')
+  } catch (e) {
+    console.error(e)
+    migrationStateStore.set('error')
+  }
```

Comment:
On implementing this, I realized that the reason I put `migrationStateStore` outside of `localSnippetsStore` is for it to be shared with `foldersStore`. Also on `localSnippetsStore`, it should be something more like: "currently queried snippets to be displayed store".

One idea for later code structure improvement: let the store be managed by a "data manager" directly, and think about putting the migration somewhere else (maybe by the "data manager" itself).